### PR TITLE
Add (toggleable) statusbar widget with word count and position sub-widgets.

### PR DESCRIPTION
**Pull request by bojjenclon**
This PR creates a new statusbar under the primary editor. This statusbar can be used to hold small widgets. The two I've created are word count and "position" (which tracks the current line number and column).

Note: You may want to ensure all of the code works the way you intend, as I haven't work with all of these libs before.

### CODE REVIEW

File: src/renderer/template/misc.css [hunk 1]
```diff
@@ -192,6 +192,7 @@
   border-left: 1px solid #d8d8d8;
   flex: 4 4 327px;
   min-width: 327px;
+  overflow: hidden;
```
Comment (fabiospampinato) [inline src/renderer/template/misc.css]: Why did you add this?

File: src/renderer/containers/main/window.ts [hunk 2]
```diff
@@ -12,6 +12,7 @@ class Window extends Container<WindowState, MainCTX> {
 
   state = {
     focus: false,
+    statusbar: false,
```
Comment (fabiospampinato) [inline src/renderer/containers/main/window.ts]: This isn't the right container for this state, you should move the code you added from this file to the "editor" container.

File: src/renderer/components/main/mainbar/statusbar_word_count.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
+
+/* IMPORT */
+
+import * as React from 'react';
+import {connect} from 'overstated';
+import Main from '@renderer/containers/main';
+
+/* STATUSBAR WORD COUNT */
+
+const WordCount = ({ content }) => {
```
Comment (fabiospampinato) [inline src/renderer/components/main/mainbar/statusbar_word_count.tsx]: Isn't there a third-party plugin that provides this to codemirror?

File: src/renderer/components/main/mainbar/statusbar_position.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    ch: 0,
+    connectedTo: null
+  }
+
+  componentDidMount () {
+    const { isEditing, getCodeMirror } = this.props;
+
+    if (isEditing) {
+      const cm = getCodeMirror(),
+        cursor = cm.getCursor();
```
Comment (fabiospampinato) [inline src/renderer/components/main/mainbar/statusbar_position.tsx]: This is not properly indented.

File: src/renderer/components/main/mainbar/statusbar_position.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+      this.setState({ line: cursor.line, ch: cursor.ch, connectedTo: cm })
+    } else if (!!connectedTo && !isEditing) {
+      connectedTo.off('cursorActivity', this.onCursorActivity.bind(this));
+
+      this.setState({ line: 0, ch: 0, connectedTo: null });
+    }
+  }
+
+  render () {
```
Comment (fabiospampinato) [inline src/renderer/components/main/mainbar/statusbar_position.tsx]: The "render" function should be last function defined by this component.

File: src/renderer/components/main/mainbar/statusbar_position.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
+
+/* IMPORT */
+
+import * as React from 'react';
+import {connect} from 'overstated';
+import Main from '@renderer/containers/main';
+
+/* STATUSBAR POSITION */
+
+class Position extends React.PureComponent<any, any> {
```
Comment (fabiospampinato) [inline src/renderer/components/main/mainbar/statusbar_position.tsx]: Isn't there a third-party plugin that provides this to codemirror?

File: src/main/windows/main.ts [hunk 7]
```diff
@@ -262,6 +262,11 @@ class Main extends Route {
             accelerator: 'CommandOrControl+Alt+F',
             click: () => this.win.webContents.send ( 'window-focus-toggle' )
           },
+          {
+            label: 'Toggle Statusbar',
+            accelerator: 'CommandOrControl+Alt+.',
```
Comment (fabiospampinato) [inline src/main/windows/main.ts]: This shortcuts sounds a bit weird to me. I would propose `CommandOrControl+Alt+S` but I'd rather use that for toggling split-view. 

Maybe it's ok to not assign a shortcut to this action as most people would rather need this or don't care about it?

File: src/renderer/components/main/mainbar/statusbar.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import WordCount from './statusbar_word_count';
+
+/* STATUSBAR */
+
+const Statusbar = ({ isStatusbar }) => {
+  if (!isStatusbar) {
+    return null;
+  }
+
+  return <div id="mainbar-statusbar" className="layout-header centerer">
```
Comment (fabiospampinato) [inline src/renderer/components/main/mainbar/statusbar.tsx]: This should be `layout-footer`, and since you're using the spacer there shouldn't be a need for `centerer`

File: src/renderer/components/main/mainbar/statusbar.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    <div className="multiple grow">
```
Comment (fabiospampinato) [inline src/renderer/components/main/mainbar/statusbar.tsx]: Maybe `multiple fluid` is more appropriate here.

File: src/main/windows/main.ts (hunk 7 above)
Comment (bojjenclon) [reply src/main/windows/main.ts]: That's actually pretty common in editors, which is how I knew about it in the first place.

File: src/renderer/components/main/mainbar/statusbar_word_count.tsx (hunk 3 above)
Comment (bojjenclon) [reply src/renderer/components/main/mainbar/statusbar_word_count.tsx]: I wasn't sure how many codemirror plugins you wanted to be reliant on in the core codebase.

File: src/renderer/template/misc.css (hunk 1 above)
Comment (bojjenclon) [reply src/renderer/template/misc.css]: Without it a second scrollbar appears at the right of the application when the statusbar is enabled. Not the editor's, but one for the entire mainbar region.

File: src/renderer/components/main/mainbar/statusbar.tsx (hunk 8 above)
Comment (bojjenclon) [reply src/renderer/components/main/mainbar/statusbar.tsx]: The default stylings provided by layout-footer look... pretty bad, for the statusbar to be frank. There'd need to be some changes to the CSS for it to look congruent.

With "layout-header centerer"
image

With "layout-footer"
image

**Comment by bojjenclon** [comment]
@fabiospampinato Some quick Google searches don't show existing plugins to CodeMirror for either of the above functions (position and word count), just threads with users explaining how one would implement said features themselves. Perhaps I'm just not familiar with the nomenclature used for CodeMirror plugins, but I don't have time to turn this into a research project so if you can point me in the direction of where I'd find such a thing I'd be happy to incorporate it, otherwise the above is the best I can provide in terms of the functionality.

**Comment by RomanKornev** [comment]
For #196

**Comment by bojjenclon** [comment]
Work has kept me insanely busy this week, but I'll be back to (hopefully) get the remaining issues here sorted out soon. Most of it has been touched already (moving around CSS selectors, changing where state is stored, reordering functions, etc.) so I'll need to quickly review the comments and see what all is missing (I'll likely just change the keybinding to the one requested by @fabiospampinato for simplicity, since ultimately its his project).

File: src/renderer/components/main/mainbar/statusbar.tsx (hunk 9 above)
Comment (bojjenclon) [reply src/renderer/components/main/mainbar/statusbar.tsx]: I tried it with fluid but I don't think that would work well for something like a statusbar. It causes all of the items to expand, whereas most of the components I'd imagine going in here (such as the included word count and position components) look better when taking up as little space as possible. Otherwise, there's just a ton of whitespace around the digits which will likely never be filled, but setting a max-width wouldn't seem appropriate either as you never know what kind of files a user may be working with.

File: src/main/windows/main.ts (hunk 7 above)
Comment (bojjenclon) [reply src/main/windows/main.ts]: I can't remember off the top of my head, I just know I've used it before. Regardless, I'm willing to change it to whatever shortcut you'd prefer. I do think that there should either be a shortcut key for it, or a way to persistently save whether its open or not between sessions. Otherwise it's a little annoying for people who _do_ want it to need to reopen it every time via the View menu.

**Comment by bojjenclon** [comment]
@fabiospampinato As long as the process (in other words, not a substantial amount of extra work) isn't all too different after the switch, sure.
